Add COB_SIGNAL_REGIME config option to control how runtime signal handler is registered

oguzcankirmemis opened the pull request:
This PR adds a config option `COB_SIGNAL_REGIME` which allows control over how the signal handler of the runtime is registered for a signal in `common.c.`/`cob_set_signal`.

# Motivation
Current implementation overwrites the signal handlers of the main program under which the runtime operates. In some environments, such as the Java Virtual Machine, this can create potential issues where signals intended for the main program are caught by the COBOL runtime, leading to corruption. For reference, Micro Focus compiler provides [a similar flag](https://www.microfocus.com/documentation/visual-cobol/vc100/DevHub/HRRTRHRTCF0J.html).

With this flag, the main program can either completely skip signal registration or skip it for specific signals it uses. Furthermore, this PR also exposes the signal handler of the runtime, allowing the main program to register it manually if desired or to combine it with its own signal handler. The latter mechanism is for instance provided by the `libjsig` [library of the Java Runtime](https://docs.oracle.com/javase/8/docs/technotes/guides/vm/signal-chaining.html).

# Implementation
- export `cob_sig_handler` function in `common.c`
- add `COB_SIGNAL_REGIME` as a configuration
- add documentation to `runtime.cfg` about `COB_SIGNAL_REGIME``
- pass HAVE_SIGNAL_H variable to tests so that the tests for COB_SIGNAL_REGIME only run when the OS provides signal handling

oguzcankirmemis committed:
feat: Add COB_SIGNAL_REGIME config

Controls how the signal handler is registered:

- 0: Register signal handler if needed (Default)
- 1: Register only if there is not previously registered signal handler ( != SIG_DFL )
- 2: Skip registering for all signals

Signed-off-by: Oguzcan Kirmemis <oguzcan.kirmemis@gmail.com>

oguzcankirmemis committed:
fix: check if COB_SIGNAL_REGIME exists

Signed-off-by: Oguzcan Kirmemis <oguzcan.kirmemis@gmail.com>

oguzcankirmemis committed:
test: add tests for COB_SIGNAL_REGIME

- pass HAVE_SIGNAL_H variable to tests to indicate if signal.h exists
- add changelog
- provide runtime config description
- export cob_sig_handler whether signal.h exists or not, if it doesn't exist,
  cob_sig_handler also does nothing
- add comments on why cob_signal_regime is under cob_settings

Signed-off-by: Oguzcan Kirmemis <oguzcan.kirmemis@gmail.com>

GitMensch commented on libcob/common.c:
```diff
@@ -1174,10 +1176,10 @@ create_dumpfile (void)
+void
 cob_sig_handler (int sig)
 {
+#ifdef	HAVE_SIGNAL_H
```
As the signals have an own check and the signal handler _should_ do something in any case, I think this ifdef needs to be dropped (one of the original reasons to be in was likely that we get unused warnings, which doesn't happen with the signal handler function being exposed now.

As it is "external ABI" now, it would be nice (not needed if you don't feel you can provide something reasonable) to add a doxygen-style function comment on this function.

oguzcankirmemis replied:
I have added the function comment. Let me know if it is an accurate description of what it does. I am not sure about removing the ifdef. Wouldn't it give a compiler error if the host does not have signal.h, since it also uses symbols from there? I thought that was the use case from the beginning. And what should the signal handler do in an environment where signal.h does not exist?

GitMensch replied:
The only stuff used without a define here would be `SIGABRT` (I guess that's defined if missing at the start) and `kill` (where the ifdef would need to be placed to (two times).

In this case the function would output "unknown signal" then does the abort handling, then `exit(sig)`.

oguzcankirmemis replied:
There are also two calls to the `signal`, which would need to be covered as well (one in the else condition of HAVE_SIGACTION ifdef, the other right after the call to `longjmp`. I can also surround them with their own ifdefs.

GitMensch replied:
sounds good - but only use "own ifdefs" if we already check them in configure (= they are part of config.h then), oitherwise use `HAVE_SIGNAL_H` for those as well (💤 )

oguzcankirmemis replied:
I have added the symbol checks for signal and kill to configure instead and used "own ifdefs" as you suggested.

GitMensch commented on libcob/common.c:
```diff
@@ -531,6 +531,7 @@ static struct config_enum beepopts[] = {{"FLASH", "1"}, {"SPEAKER", "2"}, {"FALS
 static struct config_enum timeopts[] = {{"0", "1000"}, {"1", "100"}, {"2", "10"}, {"3", "1"}, {NULL, NULL}};
 static struct config_enum syncopts[] = {{"P", "1"}, {NULL, NULL}};
 static struct config_enum varseqopts[] = {{"0", "0"}, {"1", "1"}, {"2", "2"}, {"3", "3"}, {NULL, NULL}};
+static struct config_enum sighdlrregopts[] = {{"0", "0"}, {"1", "1"}, {"2", "2"}, {NULL, NULL}};
```
If you can com up with a nice label for those values (maybe "default", "available", "ignore"?) then lpase use them in this tuple.

oguzcankirmemis replied:
I have added the labels as you suggested, but wouldn't it create confusion because we use `getenv` and directly check if `COB_SIGNAL_REGIME` is either 0,1 or 2? What I mean is for example `COB_SIGNAL_REGIME=default` wouldn't work as expected, which is the intended use case if I am understanding the configuration setup correctly.

GitMensch commented on libcob/common.c:
```diff
@@ -1519,10 +1522,17 @@ cob_set_signal (void)
 {
 #if	defined (HAVE_SIGNAL_H)
 	int k;
+	char *signal_regime;
```
please replace by

```c
	char *s = getenv ("COB_SIGNAL_REGIME");
	char signal_regime = s ? *s : '0';
```

this way you can use `signal_regime` directly in the code below and reduce duplicated code

GitMensch commented on tests/testsuite.src/run_misc.at:
```diff
@@ -15682,3 +15682,159 @@ AT_CHECK([$COBCRUN_DIRECT ./prog6], [0],
+
+
+AT_SETUP([COB_SIGNAL_REGIME={0,1} registers runtime signal handler])
+AT_KEYWORDS([runmisc COB_SIGNAL_REGIME])
```
cob_call_with_exception_check should also be included in the keywords

oguzcankirmemis replied:
Included.

GitMensch commented on tests/testsuite.src/run_misc.at:
```diff
@@ -15682,3 +15682,159 @@ AT_CHECK([$COBCRUN_DIRECT ./prog6], [0],
+
+AT_CHECK([$COMPILE caller.c], [0], [], [])
+AT_CHECK([$COMPILE_MODULE callee.cob], [0], [], [])
+AT_CHECK([COB_SIGNAL_REGIME=0 ./caller], [0], [],
```
please use the caller-prefix `$COBCRUN_DIRECT` here and below; that is f.e. used vor valgrind and other test runs

GitMensch commented on tests/testsuite.src/run_misc.at:
```diff
@@ -15682,3 +15682,159 @@ AT_CHECK([$COBCRUN_DIRECT ./prog6], [0],
+        return 0;
+    }
+
+    return 1;
```
Shouldn't that return 2 so we see the difference?

oguzcankirmemis replied:
Makes sense, thanks for pointing it out!

GitMensch commented on tests/testsuite.src/run_misc.at:
```diff
@@ -15682,3 +15682,159 @@ AT_CHECK([$COBCRUN_DIRECT ./prog6], [0],
+]])
+
+AT_CHECK([$COMPILE caller.c], [0], [], [])
+AT_CHECK([$COMPILE_MODULE callee.cob], [0], [], [])
```
Here and below, we likely need to explicit disable some runtime check with `-fno-ec=program-arg` or similar, otherwise we don't get to thje signal handler.

oguzcankirmemis replied:
As of now, the tests work without the flag, i.e. the signal handler is called. Actually, it was the easiest type of error I could find to create a segfault without any additional configuration in the compiler.

GitMensch commented on tests/atlocal_win:
```diff
@@ -167,6 +167,8 @@ if test "$GNUCOBOL_TEST_LOCAL" != "1"; then
 	done
 fi
 
+HAVE_SIGNAL_H="@HAVE_SIGNAL_H@"
```
that should likely be a constant 1 in this file (as it is not autoconfigured but copied over)

GitMensch reviewed (changes requested):
Oğuzcan that's a _very_ nice first PR!
Some minor nits below (I'll also see if we can fix the CI...)

GitMensch commented on libcob/common.c:
```diff
@@ -1535,37 +1545,57 @@ cob_set_signal (void)
+	void (*ohdlr) (int);
+
+	signal_regime = getenv ("COB_SIGNAL_REGIME");
+	if (signal && *signal_regime == '2') {
```
The change above will also fix the typo of checking signal here.

oguzcankirmemis replied:
Resolved by the comment above.

oguzcankirmemis committed:
fix: runtime.cfg of COB_SIGNAL_REGIME

- added doxygen comment for cob_sig_handler
- refactored cob_set_signal for clarity
- use COBCRUN_DIRECT to call programs in COB_SIGNAL_REGIME tests
- hardcode COB_SIGNAL_REGIME=1 in atlocal_win test config

Signed-off-by: Oguzcan Kirmemis <oguzcan.kirmemis@gmail.com>

oguzcankirmemis commented:
I am not entirely sure if the remaining failing CI checks are related to the changes in the PR. I would be glad if you could check it.

GitMensch commented:
The single macos CI failure (869) is because that environment sees the error itself before the generated code can check it; that often happens with sanitizers enabled...
The MSVC ones are rooted in vcpkg deciding it is a good idea to break msbuild projects "most use cmake in any case" (it needs a manual downgrade locally, something I had to do on appveyor months ago, I may need to "just test with CI" multiple times or setup a similar environment locally to check for the proper solution (but instead of a downgrade, one could "copy" the libxml folder to the build_windows folder then as well).
The MSYS1 one seems to build a BDB version that won't work and need to be either patched to do so after the download, or use a `CPPFLAGS=-Dsomething=else` defintion (I'd need to check, possibly with a local build as well).
MSYS2 as an install/update issue, possibly something that @ddeclerck fixed but not checked in yet.

Overall the test results look good.

GitMensch commented on libcob/common.c:
```diff
@@ -1519,10 +1538,23 @@ cob_set_signal (void)
+#endif
 
+	s = getenv ("COB_SIGNAL_REGIME");
+	signal_regime = s ? *s : '0';
```
if you move those directly up to the declaration as init code, you can even  `const char signal_regime` :-)

oguzcankirmemis replied:
That is on me, didn't know what exactly is allowed in ANSI C with variable declarations. It should be good now :)

oguzcankirmemis committed:
refactor: use own ifdefs of signal.h symbols used in cob_sig_handler

Signed-off-by: Oguzcan Kirmemis <oguzcan.kirmemis@gmail.com>

GitMensch commented on configure.ac:
```diff
@@ -711,7 +711,7 @@ AC_FUNC_VPRINTF
 	strchr strrchr strdup strstr strtol gettimeofday localeconv
 	getexecname canonicalize_file_name popen raise readlink realpath
-	setenv strcoll flockfile])
+	setenv strcoll flockfile kill signal])
```
memo to myself: there are too many functions in here (`strdup` is not used `memset` and others have to be moved to mandatory or removed in the check, `strcasecmp` is used unconditional but possibly does not exist in MSVC2005 which is still "officially supported" [maybe it is time to drop it, it also seem to be the only environment that does not provide `kill` - and not providing signal.h as well])

As noted - it is enough to check signal_h for those, no need to adjust configure (which would mean you'd need to change build_windows/config.h.in otherwise - just use `HAVE_SIGNAL_H` instead of those two checks and changing those two files).

BTW - just in case you wonder where to check things like that:
https://www.gnu.org/software/gnulib/manual/html_node/kill.html
https://en.wikipedia.org/wiki/Microsoft_Visual_C%2B%2B#Internal_version_numbering

oguzcankirmemis replied:
Nice, thanks for the tip! Reverted back to `HAVE_SIGNAL_H`.

oguzcankirmemis committed:
refactor: remove HAVE_KILL and HAVE_SIGNAL and use HAVE_SIGNAL_H directly instead

Signed-off-by: Oguzcan Kirmemis <oguzcan.kirmemis@gmail.com>

oguzcankirmemis committed:
fix: skip error sanitizers in COB_SIGNAL_REGIME tests

Signed-off-by: Oguzcan Kirmemis <oguzcan.kirmemis@gmail.com>

GitMensch reviewed (approved):
looks good, upstream asap

GitMensch commented:
@oguzcankirmemis I've just realized that we should have a NEWS entry for this (=user-visible change with as less techspeak as possible). Could you rebase and draft one at the right place (there should be more runtime settings referenced for 3.3 already?

If not then I can have a look at that as well, but that may take some days to get to.

oguzcankirmemis committed:
refactor: add COB_SIGNAL_REGIME to NEWS, adjust ChangeLogs

Signed-off-by: Oguzcan Kirmemis <oguzcan.kirmemis@gmail.com>

oguzcankirmemis commented:
I have rebased the PR and updated it with the news entry. I also adjusted the ChangeLogs slightly as well.

oguzcankirmemis commented:
There is a new failing CI (Windows MSVC), will take a look.

GitMensch commented on NEWS:
```diff
@@ -29,6 +29,10 @@ NEWS - user visible changes				-*- outline -*-
 
+** new runtime configuration COB_SIGNAL_REGIME, allows to skip registration of
+   the signal handler of GnuCOBOL. The registration can be skipped either for
+   all signals or for those that do already have a registered handler.
```
nice, maybe add something about its "user purpose" (keeping or registering own/library handlers?)

oguzcankirmemis committed:
test: skip COB_SIGNAL_REGIME tests in Windows if the compilation target is debug

Signed-off-by: Oguzcan Kirmemis <oguzcan.kirmemis@gmail.com>

GitMensch commented:
Concerning the last commit for the failing MSVC debug runtime:

Last time we tackled MSVC's runtime check @ddeclerck explicit setup its handler to not do any popups; that was back with rev5315 / 816bd2be16d88e9ea9f17f2e047fd08844d7f8c2.

The place for skipping them so far was the CI definition itself, see https://github.com/OCamlPro/gnucobol/blame/d28f9fab8e2712918ec0c15ef1d1fc476be4bcb9/.github/workflows/windows-msvc.yml#L180-L193

I'll add the skipping for that test (and also check how using the debug RTS for generated modules may be done/result in and if the "internal memory" part wouldn't only apply to the debug environment.

oguzcankirmemis commented:
Ah thank you! I didn't have time yesterday to write here, but yes, after setting up a Windows container myself locally, I have determined that the issue is that Windows CRT in Debug mode prevents the signal from reaching to our (libcob's) signal handler because it thinks it is a critical error. I also agree that "skipping" the test on the CI side makes much more sense.

Should I remove my last commit then? Or do you need something else from me for this PR?

GitMensch commented:
checked in without the last commit upstream, apart from some minor changes and the missing config/Changelog I've adjusted the runtime enum - because previously you could set signal_regime=ignore - but while being shown as expected (runtime config) it still registered everything (check of 0-2).
